reactioncommerce/reaction
# [WIP] - Marketplace

**This branch is Work in Progress**

### Completed
- Allow Sellers to add products via `Add Product`to their shop only, when they have the `createProduct` role
- Amended `shopIdAutoValue` to check loggedIn user has a shop, or fallback on parent shop
- Amended `getCurrentShop` as a start towards shop switching UI
- Amended `Product` publication to allow sub-sellers (admin) to add products to their shop
- Amended `gridControls` so that sub-sellers with `createProduct` permission can edit their products from the grid (create a new grid view for this to work)
- Allow guests to become sellers when marketplace module is enabled, `defaultSellerRoles` in Account Schema
- Added Become a seller panel in user profile page only when Marketplace plugin is enabled and `AllowGuestSeller` option is enabled
- Amended `shop/createShop` to allow guests to create a shop only once, when marketplace is enabled
- New meteor method `shop/getSellerShopId` to safely get seller shop Id, or default to parent shopId
- New `lib/api/core` for `Reaction` 
- New core functions `Reaction.addRolesToVisitors` and `Reaction.isPackageEnabled`
- Fixed `Reaction.getPackageSettings`
- Marketplace dashboard settings under `Catalog` settings with option to `allowGuestSellers` (might be better under `Settings`?)
- Fix loggedIn User dropdown perms - PR https://github.com/reactioncommerce/reaction/pull/1639
- Allow Sellers to edit their `shop settings` and change the shop's `Base Currency` https://github.com/reactioncommerce/reaction/issues/1711
- Amend `Products` Grid View and Publications to show "published" products from multiple sellers, with permissions for logged in seller to edit directly from the grid

### In Review 
1. My Shops dropdown, with Seller shop homepage - PR https://github.com/reactioncommerce/reaction/pull/1934
1. Add Shop Address, Base Unit of Measure, Timezone for Sellers - PR https://github.com/reactioncommerce/reaction/pull/1939

### In Progress
1. Stripe Connect for Sellers
1. Fix seller product image permissions to show for anonymous users
1. Image upload in product page and profile brand image, have a similar UI issue displaying the image once the upload is complete (note existing ticket on delete https://github.com/reactioncommerce/reaction/issues/1871)

### To Do
1. Shippo for Marketplace
1. Fix Dashboard Action view (sidebar) permissions for Sellers
1. Product cart flow to allow products from multiple sellers **TBD**
1. Payment flow to allow payments to multiple sellers **TBD**
1. Allow Sellers to edit their shop URL by updating Shop.domains (domain, subdomain, path based URLs)
1. Allow Tag management per shop 
1. Expand on dashboard settings, with `enableBanAccounts` see https://github.com/reactioncommerce/reaction/issues/1578, etc)

### Notes
- Introduced new client-only `SellerShops` collection - This leaves the `Shops` collection intact with the parent shop

### How to Test
1. Register as Guest
1. Go to your Profile Page
1. Scroll down and Become a Seller
1. `Shop` will replace the `Become a Seller` panel. Edit and upload a brand image (note task on image upload)
1. Click `+` at the top to `Add a Product`
1. Add information, including images
1. Click the `eye icon` to make the product public
1. Click Publish
1. Go to the homepage and see the product in the grid
